### PR TITLE
[WIP] CRUD Message Template API Routes

### LINKED ISSUE #73: CRUD Message Template Api Routes
Note: This PR does not include adding the database stuff yet - we only want to write the api first 

- [ ] Define this type (**should also include the last person who edited the message**) 
Screen Shot 2022-07-31 at 7 51 31 AM

- [ ] Write the API routes for message templates

GET a message template 
POST a message template (create)
PUT a message template (update)
DELETE a message template

I think we need to use dynamic routes for this so read up on Next.js dynamic routes 
You can use Postman to verify the API endpoints are working

### PR DESCRIPTION
**Pull request by deanframe**
Api routes for message templates

Fixes #73 

Changelist:
** We've included two different ways to make the API endpoints - we weren't sure which way is more idiomatic.**
1. Separate routes for each CRUD operation
2. A single dynamic route (`[[...id]].ts`)

We'll get rid of one and only ship whichever is better!

Tested:
- Tested via postman

### CODE REVIEW

**Comment by deanframe** [comment]
dismissing review requests, going to update to use dynamic route for GET

**Comment by deanframe** [comment]
All good now!
FYI @mike10911 @karynayen I added a new dynamic route which basically does the same thing in one endpoint instead of five separate ones. I'm not sure which is better (don't have a ton of experience) so I'm leaving both in for now!
Next.js docs on dynamic API routes: https://nextjs.org/docs/api-routes/dynamic-api-routes. I'm using an **optional catch all API route** here since I want it to handle requests that don't specify an id.

LMK if you have any questions!! The route is `[[...id]].ts`

**Comment by karynayen** [comment]
Run through of postman requests for the separate routes for each CRUD operation (NOT the dynamic route, but it's basically the same idea)
https://user-images.githubusercontent.com/37753574/183235752-45383aa0-aae9-46c3-b9da-074afeaa9288.mov

**Review by jzhzhang66 — CHANGES_REQUESTED** [review]
Woohoo!!!!!! Thanks for working on this guys<333

I think the dynamic routes way is better and more clean. 

Although I think instead of keeping everything in the [[...id]].ts and doing the id check every time we should also have an index.ts

So for the GET, PUT, and DELETE for a singlular template we can handle that logic in [...id]].ts whereas the GET and POST (when we don't need an id) we can store those in the index.ts

And then maybe we can store all the mock data in a separate file but get rid of the other endpoints (createTemplate, deleteTemplate, getAllTemplates, getTemplate, updateTemplate). This way each endpoint can handle multiple operations which I think is more standard in API's.

File: pages/api/messageTemplates/[[...id]].ts
Comment (jzhzhang66) [inline pages/api/messageTemplates/[[...id]].ts:39]: Let's move the logic for each case into its own function since we'll definitely be adding more logic as we continue to build on top of this. Plopping the applicants endpoint from the app portal here for reference! 
Screen Shot 2022-08-06 at 6 29 09 PM

File: pages/api/types.ts
Comment (jzhzhang66) [inline pages/api/types.ts:1]: In the front end we already have a Message type so can we name this one MessageTemplate or something

**Review by surbhi-gulati — COMMENTED** [review]
Looks really good! LGTM for now, and agree about preferring dynamic routes, but I'll hold off on approval till after re-review from refactoring that @jzhzhang66 suggested

**Review by jzhzhang66 — APPROVED** [review]
Looks great!!! Message is still named as Message though - could we please rename it to MessageTemplate? 

Also could we please move mockData.ts into the /lib folder and copy the types.ts contents into /lib/types.ts? We should keep all types in the same file and all the files in /api should be for a route so mockData + types should not be in there

Other than that looks great and feel free to merge!

**Comment by yanDavid21** [comment]
Great intuition y'all, while both work, dynamic routes would be the better option because a truly RESTful API has a unique URI for each resource (resource in this scenario is a message template).

Just wanted to share these resources to help in the future:
image
https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design

**Comment by deanframe** [comment]
@surbhi-gulati Could you re-review? We need another approval to merge

**Review by surbhi-gulati — APPROVED** [review]
Great work!!

**Pull request merged**
